contracts/router/Router.sol

```diff
@@ -0,0 +1,375 @@
+        uint256 _minAmountOut,
+        address[] calldata _path,
+        address[] calldata _adapters,
+        address _to
+    ) internal nonReentrant returns (uint256) {
+        require(
+            _path.length == _adapters.length + 1,
+            "Router: wrong amount of _adapters/tokens"
+        );
+        require(_to != address(0), "Router: incorrect _to address");
+        IERC20(_path[0]).safeTransfer(
+            _getDepositAddress(_path, _adapters, 0),
+            _amountIn
+        );
+
+        return _doChainedSwaps(_amountIn, _minAmountOut, _path, _adapters, _to);
+    }
+
+    /**
+     * @notice Perform a series of swaps, assuming
+     *         they have already been deposited in the first adapter
+     * @return _amount Final amount of tokens swapped
+     */
+    function _doChainedSwaps(
+        uint256 _amountIn,
+        uint256 _minAmountOut,
+        address[] calldata _path,
+        address[] calldata _adapters,
+        address _to
+    ) internal returns (uint256 _amount) {
+        for (uint256 i = 0; i < _adapters.length; i++) {
+            require(isTrustedAdapter[_adapters[i]], "Router: unknown adapter");
+        }
+        _amount = _amountIn;
+        for (uint256 i = 0; i < _adapters.length; i++) {
+            address _targetAddress = i < _adapters.length - 1
+                ? _getDepositAddress(_path, _adapters, i + 1)
+                : _to;
+            _amount = IAdapter(_adapters[i]).swap(
+                _amount,
+                _path[i],
+                _path[i + 1],
+                _targetAddress
+            );
+        }
+        require(_amount >= _minAmountOut, "Router: Insufficient output amount");
+        emit Swap(_path[0], _path[_path.length - 1], _amountIn, _amount);
+    }
+
+    // -- INTERNAL HELPERS
+
+    /**
+     * @notice Get selected adapter's deposit address
+     *
+     * @dev Return value of address(0) means that
+     *      adapter doesn't support this pair of tokens
+     */
+    function _getDepositAddress(
+        address[] calldata _path,
+        address[] calldata _adapters,
+        uint256 _index
+    ) internal view returns (address _depositAddress) {
+        _depositAddress = IAdapter(_adapters[_index]).depositAddress(
+            _path[_index],
+            _path[_index + 1]
+        );
+        require(_depositAddress != address(0), "Adapter: unknown tokens");
+    }
+
+    /**
+     * @notice Set approval for bridge to spend Router's _bridgeToken
+     *
+     * @dev This uses a finite _amount rather than UINT_MAX, so
+     *      Bridge's function redeemMax (depositMax) will be able to
+     *      pull exactly as much tokens as we need
+     */
+    function _setBridgeTokenAllowance(address _bridgeToken, uint256 _amount)
+        internal
+    {
+        IERC20 _token = IERC20(_bridgeToken);
+        uint256 allowance = _token.allowance(address(this), bridge);
+        // safeApprove should only be called when setting an initial allowance,
```

Comment:
This is called right before Bridge is supposed to pull as much tokens from Router, as it is allowed. This is why we use finite allowance instead of UINT_MAX. 

If Router happens to have some extra tokens lying around, these tokens should absolutely **not** be pulled - this is why we need to set allowance exactly to `_amount`, instead of using `_token.safeIncreaseAllowance(bridge, _amount)`